staging/src/k8s.io/component-helpers/node/util/nodeaddresses.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"net"
+
+	v1 "k8s.io/api/core/v1"
+	netutils "k8s.io/utils/net"
+)
+
+// PreferNodeIP filters node addresses to prefer a specific node IP or address
+// family.
+//
+// If nodeIP is either '0.0.0.0' or '::' it is taken to represent any address of
+// that address family: IPv4 or IPv6. i.e. if nodeIP is '0.0.0.0' we will return
+// node addresses sorted such that all IPv4 addresses are listed before IPv6
+// addresses.
+//
+// If nodeIP is a specific IP, either IPv4 or IPv6, we will return node
+// addresses filtered such that:
+// * Any address matching nodeIP will be listed first.
+// * If nodeIP matches an address of a particular type (internal or external),
+//   that will be the *only* address of that type returned.
+// * All remaining addresses are listed after.
+func PreferNodeIP(nodeIP net.IP, cloudNodeAddresses []v1.NodeAddress) ([]v1.NodeAddress, error) {
```

Comment:
> It's now called by both kubelet and cloud-provider, so I don't think we can move it into cloud provider. 

But it's only the part of the kubelet dealing with cloud node addresses right? Or does it also pertain to --node-ip without the cloud provider?

Reply:
> It's the code in kubelet which re-orders node addresses when not using an external cloud provider.

When not using an external cloud provider, but still using the internal/in-tree one though right? k8s.io/cloud-provider can still contain helper functions used by in-tree cloud provider code

Reply:
If the usage of PreferNodeIP can be used when kubelet does not set the --cloud-provider flag, then k8s.io/component-helpers makes sense to me. If it's exclusively used when cloud provider is enabled (external or internal), then k8s.io/cloud-provider probably makes more sense since the kubelet will (eventually) not use that code at some point

Reply:
@andrewsykim Sorry, I hadn't understood. Yes, this code will only be called when there is *some* cloud provider, either internal or external.

@danwinship @mrunalp Any reason I should not respin this and move the node address code to the suggested location in cloud provider instead? We already have deps on cloud-provider and cloud-provider/api in this file, so I assume this wouldn't be problematic?

Reply:
I have no opinion on where it belongs...